text/0000-async-work-service.md

```diff
@@ -0,0 +1,322 @@
+
+
+
+### Requirements
+
+| ID   | Statement                                                    | Source / Pain Point | Priority  |
+| ---- | ------------------------------------------------------------ | ------------------- | --------- |
+| 1    | The service interface should be re-usable by any other CppMicroServices service or application. | PP1                 | Must Have |
+| 2    | The service interface should be decoupled from CppMicroServices and all compendium services. | PP2                 | Must Have |
+| 3    | CppMicroServices service based design                        | OSGi compliance     | Must Have |
+| 4    | Allow users to block on the asynchronous task until it has finished. | PP3                 | Must Have |
+| 5    | Allow users to receive a failure from the asynchronous task. | PP3                 | Must Have |
+
+
+
+## Detailed design
+
+> This is the bulk of the RFC.
+
+> Explain the design in enough detail for somebody
+familiar with the framework to understand, and for somebody familiar with the
+implementation to implement. This should get into specifics and corner-cases,
+and include examples of how the feature is used. Any new terminology should be
+defined here.
+
+
+
+### API
+
+```c++
+namespace cppmicroservices { 
+  namespace async {
+    /**
+     *
+     */
+    class UnboundedAsyncTask {
+    public:
+      virtual ~UnboundedAsyncTask() noexcept = default;
+      
+      /**
+       * Run a std::function<void()> object asynchronously on another thread.
+       *
+       * @param task A std::function<void()> object representing a Callable target.
+       * @return A std::future<void> object to wait on the result of task.
+       * @note There is no guarantee made by this interface that the 
+       * std::future<void> returned will automatically block when it goes out of scope. 
+       * It may or may not depending on the implementation of this method.
+       */
+      virtual std::future<void> post(std::function<void()>&& task) = 0;
+      
+      /**
+       * Run a std::packaged_task<void()> asynchronously on another thread.
+       * The std::future<void> associated with the std::packaaged_task<void()>
+       * task object will contain the result from the task object.
+       *
+       * @param task A std::packaged_task<void()> wrapping a Callable target
+       * to execute asynchronously.
+       *
+       * @note The caller is required to manage the std::future<void> associated
+       * with the std::packaged_task<void()> in order to wait on the async task.
+       * 
+       */
+      virtual void post(std::packaged_task<void()>&& task) = 0;
```

Comment:
this API may not be needed; the `std::function` version of `post` could be used to accomplish the same thing.

Reply:
From a prototype I created to make sure that the requirements stated for the interface work for our usage in CppMicroServices, I determined that the version of post which takes a std::function<void()>&& is actually the version which is not necessary. The design of DeclarativeServices necessitates the usage of packaged_task and ConfigurationAdmin can be quickly refactored to make it work with packaged_task in all use cases.

The packaged_task version, in my opinion, is better because packaged_task objects already have an associated future. If a user of the service were to use the std::function<void()>&& version of post, then it is ambiguous as to how to create a future for the asynchronous task. Should the implementer use std::async which returns a future? What if the implementer of the service made the post function "inline"; how is the future, that must be returned, constructed?

Only having this version and requiring the usage of packaged_task ensures the existence of a valid future that the user can manage and maintain in every use case.